include/aws/s3/s3_client.h

```diff
@@ -653,12 +675,121 @@ struct aws_s3_meta_request *aws_s3_meta_request_acquire(struct aws_s3_meta_reque
 AWS_S3_API
 struct aws_s3_meta_request *aws_s3_meta_request_release(struct aws_s3_meta_request *meta_request);
 
+/**
+ * Initialize the configuration for a default S3 signing.
+ */
 AWS_S3_API
 void aws_s3_init_default_signing_config(
     struct aws_signing_config_aws *signing_config,
     const struct aws_byte_cursor region,
     struct aws_credentials_provider *credentials_provider);
 
+/**
+ * Add a reference, keeping this object alive.
+ * The reference must be released when you are done with it, or it's memory will never be cleaned up.
+ * Always returns the same pointer that was passed in.
+ */
+AWS_S3_API
+struct aws_s3_request_metrics *aws_s3_request_metrics_acquire(struct aws_s3_request_metrics *metrics);
+
+/**
+ * Release a reference.
+ * When the reference count drops to 0, this object will be cleaned up.
+ * It's OK to pass in NULL (nothing happens).
+ * Always returns NULL.
+ */
+AWS_S3_API
+struct aws_s3_request_metrics *aws_s3_request_metrics_release(struct aws_s3_request_metrics *metrics);
+
+/*************************************  Getters for s3 request metrics ************************************************/
+/* Get the request ID from aws_s3_request_metrics, AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE will be raised when data
+ * unavailable */
+AWS_S3_API
+int aws_s3_request_metrics_get_request_id(
+    const struct aws_s3_request_metrics *metrics,
+    struct aws_byte_cursor *out_request_id);
+
+/* Ge the start time from aws_s3_request_metrics, which is when S3 client prepare the request to be sent. Always be
+ * available. Timestamp are from `aws_high_res_clock_get_ticks`  */
+AWS_S3_API
+void aws_s3_request_metrics_get_start_timestamp_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *out_start_time);
+
+/* Ge the end time from aws_s3_request_metrics. Always be available */
+AWS_S3_API
+void aws_s3_request_metrics_get_end_timestamp_ns(const struct aws_s3_request_metrics *metrics, uint64_t *out_end_time);
+
+/* Ge the total duration time from aws_s3_request_metrics. Always be available */
+AWS_S3_API
+void aws_s3_request_metrics_get_total_duration_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *out_total_duration);
+
+/* Ge the start time stamp when request was sent to the network channel from aws_s3_request_metrics.
+ * AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE will be raised if the request was end before it gets sent. */
+AWS_S3_API
+int aws_s3_request_metrics_get_send_start_timestamp_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *out_send_start_time);
+
+AWS_S3_API
+int aws_s3_request_metrics_get_send_end_timestamp_ns(
```

Comment:
I think we should add docs for all the functions and mention if they will be always available or not. Since the actual struct is now private, we should copy paste the documentation here as well.